### PR TITLE
Add recent work widget [SCI-4400][SCI-4401][SCI-4402][SCI-4403]

### PR DESCRIPTION
**Pull request by aignatov-bio**
Jira ticket: [SCI-4400](https://biosistemika.atlassian.net/browse/SCI-4400)
Jira ticket: [SCI-4401](https://biosistemika.atlassian.net/browse/SCI-4401)
Jira ticket: [SCI-4402](https://biosistemika.atlassian.net/browse/SCI-4402)
Jira ticket: [SCI-4403](https://biosistemika.atlassian.net/browse/SCI-4403)

### What was done
Add recent work widget

### CODE REVIEW

File: app/assets/javascripts/dashboard/recent_work.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+      if ($('.recent-work-widget').length) {
+        initNavbar();
+        initRecentWork();
+      }
+    }
+  };
+}());
+
+$(document).on('turbolinks:load', function() {
+  DasboardRecentWorkWidget.init();
```
Comment (okriuchykhin) [inline app/assets/javascripts/dashboard/recent_work.js]: What happens on navigation return?

File: app/assets/javascripts/dashboard/recent_work.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  function initRecentWork() {
+    var container = $('.recent-work-container');
+    $('.recent-work-container').empty();
+    $.get(container.data('url'), {
+      mode: $('.recent-work-navbar .active').data('mode')
+    }, function(result) {
+      if (result.length) {
+        renderRecentWorkItem(result, container);
+      } else {
+        container.append(`<div class="no-results">
```
Comment (okriuchykhin) [inline app/assets/javascripts/dashboard/recent_work.js]: Move it to html.erb

File: app/assets/javascripts/dashboard/recent_work.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
+/* global I18n PerfectSb*/
+/* eslint-disable no-param-reassign */
+
+var DasboardRecentWorkWidget = (function() {
+  function renderRecentWorkItem(data, container) {
+    $.each(data, (i, item) => {
+      var recentWorkItem = `<a href="${item.url}" class="recent-work-item">
```
Comment (okriuchykhin) [inline app/assets/javascripts/dashboard/recent_work.js]: Move it to html.erb

File: app/assets/javascripts/dashboard/recent_work.js (hunk 1 above)
Comment (aignatov-bio) [reply app/assets/javascripts/dashboard/recent_work.js]: `tubolinks:load` correctly trigger on restoration visits.

Comment (okriuchykhin) [reply app/assets/javascripts/dashboard/recent_work.js]: yes, all over the app :man_facepalming:

Comment (aignatov-bio) [reply app/assets/javascripts/dashboard/recent_work.js]: It will run only if widget exist

File: app/services/dashboard/recent_work_service.rb [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+                          subject_type = 'Repository'
+                          AND subject_id = repositories.id
+                        LEFT JOIN reports ON subject_type = 'Report'
+                          AND subject_id = reports.id
+                      ")
+                   .where("
+                        (projects.archived != 'true' OR projects.archived IS NULL)
+                        AND (experiments.archived != 'true' OR experiments.archived IS NULL)
+                        AND (my_modules.archived != 'true' OR my_modules.archived IS NULL)
+                        AND (protocols.protocol_type != 4 OR protocols.protocol_type IS NULL)
```
Comment (okriuchykhin) [inline app/services/dashboard/recent_work_service.rb]: All these are NOT NULL

File: app/services/dashboard/recent_work_service.rb [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+                        reports.project_id as report_project_id,
+                        subject_id,
+                        subject_type,
+                        last_change
+                      ')
+
+      query_filter = []
+      if %w(all projects).include? @mode
+        query_filter.push("group_id LIKE '%tsk%' OR group_id LIKE '%exp%' OR group_id LIKE '%pro%'")
```
Comment (okriuchykhin) [inline app/services/dashboard/recent_work_service.rb]: Why do you search in the middle of the string?

File: app/services/dashboard/recent_work_service.rb [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+                          OR experiments.id = my_modules.experiment_id
+                        LEFT JOIN projects ON
+                          (subject_type = 'Project' AND subject_id = projects.id)
+                          OR projects.id = experiments.project_id
+                        LEFT JOIN repositories ON
+                          subject_type = 'Repository'
+                          AND subject_id = repositories.id
+                        LEFT JOIN reports ON subject_type = 'Report'
+                          AND subject_id = reports.id
+                      ")
```
Comment (okriuchykhin) [inline app/services/dashboard/recent_work_service.rb]: Move it to scopes

File: app/services/dashboard/recent_work_service.rb (hunk 4 above)
Comment (okriuchykhin) [reply app/services/dashboard/recent_work_service.rb]: query = query.where('projects.archived IS NOT TRUE')
                   .where('experiments.archived IS NOT TRUE')
                   .where('my_modules.archived IS NOT TRUE')
                   .where('protocols.protocol_type != ? OR protocols.protocol_type IS NULL',
                          Protocol.protocol_types[:in_repository_archived])

File: app/services/dashboard/recent_work_service.rb [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+                          CONCAT(\'pro\', projects.id)
+                        WHEN protocols.id IS NOT NULL THEN
+                          CONCAT(\'prt\', protocols.id)
+                        WHEN repositories.id IS NOT NULL THEN
+                          CONCAT(\'inv\', repositories.id)
+                        WHEN reports.id IS NOT NULL THEN
+                          CONCAT(\'rpt\', reports.id)
+                        END as group_id,
+
+                        CASE
```
Comment (okriuchykhin) [inline app/services/dashboard/recent_work_service.rb]: Use COALESCE

File: app/services/dashboard/recent_work_service.rb [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+                          projects.name
+                        WHEN protocols.name IS NOT NULL THEN
+                          protocols.name
+                        WHEN repositories.name IS NOT NULL THEN
+                          repositories.name
+                        WHEN reports.name IS NOT NULL THEN
+                          reports.name
+                        END as name,
+
+                        reports.project_id as report_project_id,
```
Comment (okriuchykhin) [inline app/services/dashboard/recent_work_service.rb]: AS

File: app/services/dashboard/recent_work_service.rb [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+                          experiments.name
+                        WHEN projects.name IS NOT NULL THEN
+                          projects.name
+                        WHEN protocols.name IS NOT NULL THEN
+                          protocols.name
+                        WHEN repositories.name IS NOT NULL THEN
+                          repositories.name
+                        WHEN reports.name IS NOT NULL THEN
+                          reports.name
+                        END as name,
```
Comment (okriuchykhin) [inline app/services/dashboard/recent_work_service.rb]: AS

File: app/services/dashboard/recent_work_service.rb [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    def initialize(user, team, mode)
+      @user = user
+      @team = team
+      @mode = mode
+    end
+
+    def call
+      visible_projects = Project.viewable_by_user(@user, @team)
+
+      activities = Activity.where("(values #>> '{message_items, user, id}') :: BIGINT = ?", @user.id)
```
Comment (okriuchykhin) [inline app/services/dashboard/recent_work_service.rb]: Why do you search user in JSON and do not use owner_id?

File: app/models/activity.rb [hunk 11]
```diff
@@ -21,6 +21,32 @@ class Activity < ApplicationRecord
   validates :subject_type, inclusion: { in: Extends::ACTIVITY_SUBJECT_TYPES,
                                         allow_blank: true }
 
+  scope :subjects_joins, lambda {
```
Comment (okriuchykhin) [inline app/models/activity.rb]: split it and give meaningful names

**Pull request merged**